lib/modules.nix

```diff
@@ -463,11 +463,19 @@ rec {
       # yield a value computed from the definitions
       value = if opt ? apply then opt.apply res.mergedValue else res.mergedValue;
 
-      warnDeprecation =
-        if opt.type.deprecationMessage == null then id
-        else warn "The type `types.${opt.type.name}' of option `${showOption loc}' defined in ${showFiles opt.declarations} is deprecated. ${opt.type.deprecationMessage}";
+      # Issue deprecation warnings recursively over all subtypes of the given type
+      # But don't recurse if the type was already visited before, there are recursive types after all
+      # Returns the new set of visited type names
+      recursiveWarn = visited: type:
+        let
+          maybeWarn = res:
+            if type.deprecationMessage == null then res
+            else warn "The type `types.${type.name}' of option `${showOption loc}' defined in ${showFiles opt.declarations} is deprecated. ${type.deprecationMessage}" res;
+        in
+          if visited ? ${type.name} then visited
```

Comment:
Oh, so the check is just done over the name of the type, since deprecation is once per “class” of the type, not of the actual value of the type (which might contain subtypes).

This will not work if two types should ever share the same name, but is probably good enough for this check.

Maybe a comment that states this assumption/restriction is in order.

Reply:
Made the comment a bit clearer regarding this